electrical/breakbeam.py

```diff
@@ -0,0 +1,90 @@
+import RPi.GPIO as GPIO
+import time
+from enum import Enum
+
+
+class Breakbeam:
+    ACCEPTABLE_TIME = 3 # time to wait to confirm a full break
+    class BEAM_PINS(Enum):
+        HALF1 = 17
+        HALF2 = 22
+        FULL1 = 23
+        FULL2 = 24
+
+    def __init__(self):
+        """
+        Initializes each sensor to detect changes. Initializes beam_broken to False.
+        Creates an empty list for the beams that are broken.
+        """
+        for BEAM_PIN in Breakbeam.BEAM_PINS:
+            GPIO.setmode(GPIO.BCM)
+            GPIO.setup(BEAM_PIN.value, GPIO.IN, pull_up_down=GPIO.PUD_UP)
+            GPIO.add_event_detect(BEAM_PIN.value, GPIO.BOTH, callback=self.break_beam_callback)
+        self.blocked_sensors = set()
+    
+    def break_beam_callback(self, channel):
+        """
+        Returns True if the beam at channel has been broken.
+        Parameters: channel - the pin used by this sensor
+        """
+        if not GPIO.input(channel):
+            self.timer(channel)
+            self.check_half()
+            self.check_full()
+        # print(self.check_half())
+        # print(self.check_full())
+    
+    def timer(self, channel):
+        """
+        Returns the pin number of this sensor.
+        Gets the current time and calculates the time that has passed, then compares
+        this to the ACCEPTABLE_TIME to determine whether the broken beam was just
+        tripped once or if it is fully broken. If fully broken, add it to the set
+        blocked_sensors.
+        Parameters: channel - the pin used by this sensor
+        """
+        start = time.time()
+        # loops until ACCEPTABLE_TIME and then goes into the else statement once
+        while time.time() - start < Breakbeam.ACCEPTABLE_TIME:
+            if(GPIO.input(channel)):
```

Comment:
what is this if statement checking? is this just an ece thing?

Reply:
I believe it is an ece thing, where the GPIO.input(channel) retakes the reading when we call it but I'm not sure if that's how it works. We were discussing whether storing GPIO.input(channel) in a variable would make it store the reading for only that one time, so just to be safe we removed the variable and just use GPIO.input(channel).

Reply:
But the if statement checks whether the sensor becomes unblocked and exits the while-else loop if this is the case.